src/ratingservice/main.py

```diff
@@ -13,103 +13,180 @@
+
+    Args:
+        eid (string): the entity id.
+
+    Returns:
+        HTTP status 200 and Json payload { 'id': (string), 'rating': (number), 'votes': (int) }
+        HTTP status 400 when eid is is missing or invalid
+        HTTP status 404 when rating for eid cannot be found
+        HTTP status 500 when there is an error querying DB
+    '''
+
+    if eid == None or eid == "":
+        return makeError(400, "malformed entity id")
+
+    conn = getConnection()
+    if conn == None:
+        return makeError(500, 'failed to connect to DB')
     try:
-        products = read_products()
-        conn = connpool.getconn()
         with conn.cursor() as cursor:
-            cursor.execute("SELECT EXISTS(SELECT * FROM information_schema.tables WHERE table_name = 'ratings');")
+            cursor.execute(
+                "SELECT rating, votes FROM ratings WHERE eid=%s", (eid,))
             result = cursor.fetchone()
-            if not result[0]:
-                cursor.execute("CREATE TABLE ratings (id SERIAL PRIMARY KEY, product_id varchar(20) NOT NULL, score int DEFAULT 0);")
-                for product in products:
-                    cursor.execute("INSERT INTO ratings (product_id, score) VALUES ('{}', 5);".format(product))
         conn.commit()
+        if result != None:
+            return makeResult({
+                'id': eid,
+                'rating': result[0],
+                'votes': result[1]
+            })
+        else:
+            return makeError(404, "invalid entity id")
+    except:
+        resp = makeError(500, 'DB error')
     finally:
-        connpool.putconn(conn)
+        db_connection_pool.putconn(conn)
 
-app = Flask(__name__)
-connpool = initConnection()
-populate_database()
-
-# get rating of a product
-@app.route('/rating/<id>', methods=['GET'])
-def getRating(id):
-    conn = connpool.getconn()
-    resp = None
+
+@app.route('/rating', methods=['POST'])
+def postRating():
+    '''Adds new vote for entity's rating.
+
+    Args:
+        Json payload {'id': (string), 'rating': (integer) }
+
+    Returns:
+        HTTP status 200 and empty Json payload { }
+        HTTP status 400 when payload is malformed (e.g. missing expected field)
+        HTTP status 400 when eid is missing or invalid or rating is missing, invalid or out of [1..5] range
+        HTTP status 404 when rating for eid cannot be reported
+        HTTP status 500 when there is an error querying DB
+    '''
+
+    data = request.get_json()
+    if data == None:
+        return makeError(400, "missing json payload")
+
+    eid = str(data['id'])
+    if eid == None or eid == "":
+        return makeError(400, "malformed entity id")
+    rating_str = str(data['rating'])
+    if rating_str == None or rating_str == "":
+        return makeError(400, "malformed rating")
+    if not rating_str.isdigit():
+        return makeError(400, "rating should be non negative number")
+    rating = int(rating_str)
```

Comment:
You are assuming here that everything that passes `isdigit` will be parsable by `int()`. If the string is something like `"²"`, a ValueError would be thrown here

in general, it would be better to put more of this parsing code in try blocks, and return errors when an exception is thrown, rather than trying to validate types manually

Reply:
refactored